contributors/design-proposals/storage/container-storage-interface.md

```diff
@@ -0,0 +1,434 @@
+// VolumeAttachment objects are non-namespaced.
+type VolumeAttachment struct {
+  metav1.TypeMeta `json:",inline"`
+  metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+  // Specification of the desired attach/detach volume behavior.
+  Spec VolumeAttachmentSpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+
+  // Most recently observed status of the VolumeAttachment request.
+  // This data may not be up to date.
+  Status VolumeAttachmentStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+
+type VolumeAttachmentSpec struct {
+  // Attacher indicates the name of the volume plugin that MUST handle this
+  // request. This is the name returned by GetPluginName().
+  Attacher string `json:"attacher" protobuf:"bytes,1,opt,name=attacher"`
+
+  // Volume represents the volume that should be attached.
+  Volume AttachedVolumeSource `json:",inline" protobuf:"bytes,2,opt,name=volume"`
+
+  // NodeName indicates the node that the volume should be attached to.
+  NodeName string `json:"nodeName,omitempty" protobuf:"bytes,3,opt,name=nodeName"`
+}
+
+// VolumeAttachmentSource represents a volume that should be attached.
+// Right now only PersistenVolumes can be attached via external attacher,
+// in future we may allow also inline volumes in pods.
+// Exactly one member can be set.
+type AttachedVolumeSource struct {
+    // Name of the persistent volume to attach.
+    // +optional
+    PersistentVolumeName *string `json:"persistentVolumeName,omitempty" protobuf:"bytes,1,opt,name=persistentVolumeName"`
+
+    // Placeholder for *VolumeSource to accommodate inline volumes in pods.
+}
+
+type VolumeAttachmentStatus struct {
+  // IsAttached indicates the volume is successfully attached.
+  // This field must only be set by the entity completing the attach
+  // operation, i.e. the external-attacher.
+  IsAttached bool `json:"isAttached" protobuf:"varint,1,opt,name=isAttached"`
+
+  // Upon successful attach, this field is updated with any returned
+  // information that must be passed into subsequent WaitForAttach or
+  // mount calls.
+  // This field must only be set by the entity completing the attach
+  // operation, i.e. the external-attacher.
+  AttachmentMetadata map[string]string `json:"attachmentMetadata,omitempty" protobuf:"bytes,2,rep,name=attachmentMetadata"`
+
+  // Last error encountered during attach operation, if any.
+  // +optional
+  // This field must only be set by the entity completing the attach
+  // operation, i.e. the external-attacher.
+  AttachError VolumeError `json:"attachError,omitempty" protobuf:"bytes,3,opt,name=attachError,casttype=VolumeError"`
+
+  // Last error encountered during detach operation, if any.
+  // +optional
+  // This field must only be set by the entity completing the detach
+  // operation, i.e. the external-attacher.
+  DetachError VolumeError `json:"detachError,omitempty" protobuf:"bytes,3,opt,name=detachError,casttype=VolumeError"`
```

Comment:
optional structs must be pointers to be able to be omitted when empty

Reply:
fixed, however I can see an optional struct that's not a pointer:

```
type PersistentVolume struct {
       ...
	// +optional
	Status PersistentVolumeStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
```

I will use the same approach for `VolumeAttachmentStatus`, we allow it to be created without status (which implies `attached == false`).